src/Raven.Server/Documents/Replication/OutgoingReplicationHandler.cs

```diff
@@ -463,15 +467,22 @@ private void WriteHeaderToRemotePeer()
             using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext documentsContext))
             using (var writer = new BlittableJsonTextWriter(documentsContext, _stream))
             {
-                documentsContext.Write(writer, new DynamicJsonValue
+                WriteHeaderInternal(documentsContext, writer, TcpConnectionHeaderMessage.ReplicationTcpVersion);
+                var destinationProtocol = ReadHeaderResponseAndThrowIfUnAuthorized(documentsContext, writer);
+                if (destinationProtocol != TcpConnectionHeaderMessage.ReplicationTcpVersion)
```

Comment:
Here we need to do this in a loop.
The logic goes like this:

10 Src sends its highest version. If dest accepts it, great. 
20 If not, the src sends its highest version. If the dest knows of it, great.
30 If not, src sends its next highest version, repeat until there is a version that both of them understand.

The idea is that we'll use the highest version that both endpoints understand.
Src will try them in order.
Dest will check if it understand the protocol version and _always_ reply with its own preferred version.

Reply:
Yes, this needs to change according to our talk yesterday.
This ping pong will be handled at the same level on the accepting side as now but each sender will have to modify the logic to support this. This will actually not break anything since we already allow 3 streaks and since we don't have more than 3 version for any of our protocol increasing the retry won't break anything :)
I'll add the agreed upon version with the TCP stream and each logic will have to behave accordingly.